File: src/backtester.ts
/**
 * backtester.ts - 回测引擎
 *
 * 职责：用历史 K 线和策略信号模拟真实交易，计算收益和风险指标。
 *
 * 交易规则：
 *   - 初始资金 $10,000，持仓 0
 *   - 买入信号 → 全部资金买入（按当前收盘价）
 *   - 卖出信号 → 全部持仓卖出
 *   - 最后如果还持有，按最后一根收盘价强制平仓
 */

import { StrategyCandle } from './strategy.js';

/** 单笔交易记录 */
export interface Trade {
  time: number;
  type: 'buy' | 'sell';
  price: number;
  amount: number;       // 币数量
  value: number;        // 交易金额
  balanceAfter: number; // 交易后现金余额
}

/** 回测结果 */
export interface BacktestResult {
  initialCapital: number;
  finalCapital: number;
  totalReturn: number;
  maxDrawdown: number;
  trades: Trade[];
  totalTrades: number;
  winningTrades: number;
  buyAndHoldReturn: number;
}

/**
 * 执行回测
 */
export function backtest(
  candles: StrategyCandle[],
  initialCapital: number = 10000
): BacktestResult {
  let cash = initialCapital;
  let coins = 0;
  const trades: Trade[] = [];

  // 记录每一步的资产价值，用于计算回撤
  const equityHistory: number[] = [];

  for (let i = 0; i < candles.length; i++) {
    const c = candles[i];
    const price = c.close;

    // === 处理买入信号 ===
    if (c.signal === 'buy' && coins === 0 && cash > 0) {
      coins = cash / price;
      cash = 0;

      trades.push({
        time: c.time,
        type: 'buy',
        price,
        amount: coins,
        value: coins * price,
        balanceAfter: 0,
      });
    }

    // === 处理卖出信号 ===
    if (c.signal === 'sell' && coins > 0) {
      const soldAmount = coins;   // 先保存数量
      cash = coins * price;
      coins = 0;

      trades.push({
        time: c.time,
        type: 'sell',
        price,
        amount: soldAmount,        // 使用保存的数量
        value: cash,
        balanceAfter: cash,
      });
    }

    // === 记录当前资产（现金 + 持仓市值） ===
    equityHistory.push(cash + coins * price);
  }

  // === 最后如果还持仓，强制平仓 ===
  if (coins > 0) {
    const lastPrice = candles[candles.length - 1].close;
    const finalAmount = coins;
    cash = coins * lastPrice;
    coins = 0;

    trades.push({
      time: candles[candles.length - 1].time,
      type: 'sell',
      price: lastPrice,
      amount: finalAmount,
      value: cash,
      balanceAfter: cash,
    });
  }

  // === 计算总收益 ===
  const finalCapital = cash;
  const totalReturn = (finalCapital - initialCapital) / initialCapital;

  // === 计算最大回撤 ===
  let maxDrawdown = 0;
  let peak = initialCapital;
  for (const equity of equityHistory) {
    if (equity > peak) peak = equity;
    const drawdown = (peak - equity) / peak;
    if (drawdown > maxDrawdown) maxDrawdown = drawdown;
  }

  // === 买入持有收益（对照组） ===
  const firstPrice = candles[0].close;
  const lastPrice = candles[candles.length - 1].close;
  const bhCoins = initialCapital / firstPrice;
  const bhFinal = bhCoins * lastPrice;
  const buyAndHoldReturn = (bhFinal - initialCapital) / initialCapital;

  // === 胜率 ===
  let winningTrades = 0;
  // 每对买卖（i 是 buy, i+1 是 sell）比较盈亏
  for (let i = 0; i < trades.length - 1; i += 2) {
    if (trades[i].type === 'buy' && trades[i + 1].type === 'sell') {
      if (trades[i + 1].value > trades[i].value) {
        winningTrades++;
      }
    }
  }

  return {
    initialCapital,
    finalCapital,
    totalReturn,
    maxDrawdown,
    trades,
    totalTrades: trades.length,
    winningTrades,
    buyAndHoldReturn,
  };
}

File: src/backtester.py
"""backtester.py — 回测引擎

模拟真实交易：资金管理、滑点、手续费、最大回撤计算。
"""
from dataclasses import dataclass, field
import pandas as pd


@dataclass
class Trade:
    time: int          # 时间戳 ms
    type: str          # "buy" | "sell"
    price: float
    amount: float      # 币数量
    value: float       # 交易金额
    balance_after: float


@dataclass
class BacktestResult:
    initial_capital: float
    final_capital: float
    total_return: float
    max_drawdown: float
    sharpe_ratio: float
    trades: list[Trade] = field(default_factory=list)
    buy_hold_return: float = 0.0
    sentiment_avg: float = 0.0  # 持仓期间平均情绪


def backtest(
    df: pd.DataFrame,
    initial_capital: float = 10000,
    fee_rate: float = 0.001,  # 0.1% 手续费
    slippage: float = 0.0005,  # 0.05% 滑点
) -> BacktestResult:
    """执行回测。

    df 需包含: close, signal 列
    """
    cash = initial_capital
    coins = 0.0
    trades: list[Trade] = []
    equity_curve: list[float] = []
    sentiment_held: list[float] = []

    for i in range(len(df)):
        row = df.iloc[i]
        price = row["close"]
        signal = row["signal"]
        sentiment = row.get("sentiment", 0)

        # 执行价格含滑点
        exec_price = price * (1 + slippage) if signal == "buy" else price * (1 - slippage)

        # —— 买入 ——
        if signal == "buy" and coins == 0 and cash > 0:
            gross = cash * (1 - fee_rate)
            coins = gross / exec_price
            cash = 0
            trades.append(Trade(
                time=row["time"],
                type="buy",
                price=exec_price,
                amount=coins,
                value=coins * exec_price,
                balance_after=0,
            ))

        # —— 卖出 ——
        if signal == "sell" and coins > 0:
            gross = coins * exec_price
            cash = gross * (1 - fee_rate)
            coins = 0
            trades.append(Trade(
                time=row["time"],
                type="sell",
                price=exec_price,
                amount=trades[-1].amount if trades else 0,
                value=gross,
                balance_after=cash,
            ))

        # 记录持仓时的情绪
        if coins > 0 and not pd.isna(sentiment):
            sentiment_held.append(sentiment)

        equity_curve.append(cash + coins * price)

    # —— 强制平仓 ——
    if coins > 0:
        last_price = df["close"].iloc[-1] * (1 - slippage)
        cash = coins * last_price * (1 - fee_rate)
        trades.append(Trade(
            time=df["time"].iloc[-1],
            type="sell",
            price=last_price,
            amount=coins,
            value=coins * last_price,
            balance_after=cash,
        ))
        coins = 0

    final_capital = cash
    total_return = (final_capital - initial_capital) / initial_capital

    # 最大回撤
    peak = initial_capital
    max_dd = 0.0
    for eq in equity_curve:
        if eq > peak:
            peak = eq
        dd = (peak - eq) / peak
        if dd > max_dd:
            max_dd = dd

    # 夏普比率（简化，假设无风险利率=0）
    if len(equity_curve) > 1:
        returns = pd.Series(equity_curve).pct_change().dropna()
        sharpe = (returns.mean() / returns.std()) * (252 ** 0.5) if returns.std() > 0 else 0
    else:
        sharpe = 0

    # 买入持有
    first = df["close"].iloc[0]
    last = df["close"].iloc[-1]
    bh = (initial_capital / first) * last
    bh_return = (bh - initial_capital) / initial_capital

    return BacktestResult(
        initial_capital=initial_capital,
        final_capital=final_capital,
        total_return=total_return,
        max_drawdown=max_dd,
        sharpe_ratio=round(sharpe, 4),
        trades=trades,
        buy_hold_return=bh_return,
        sentiment_avg=round(sum(sentiment_held) / len(sentiment_held), 4) if sentiment_held else 0,
    )
